### PR TITLE
PERF: MultiIndex.equals

### LINKED ISSUE #43549: PERF: MultiIndex equals method got 20x slower
### 

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this issue exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this issue exists on the master branch of pandas.

### Reproducible Example

from dateutil import tz
import pandas as pd
from time import time

dates = pd.date_range('2010-01-01', periods=1000, tz=tz.tzutc())
index = pd.MultiIndex.from_product([range(100), dates])
index2 = index.copy()
start_time = time()
index.equals(index2)
used_time = time() - start_time
print("Took", int(used_time*1000), "milliseconds")

### Installed Versions

<details>

commit           : 73c68257545b5f8530b7044f56647bd2db92e2ba
python           : 3.9.6.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 20.5.0
Version          : Darwin Kernel Version 20.5.0: Sat May  8 05:10:33 PDT 2021; root:xnu-7195.121.3~9/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : en_US.UTF-8
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.3.3
numpy            : 1.21.2
pytz             : 2021.1
dateutil         : 2.8.2
pip              : 21.2.4
setuptools       : 57.4.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : None

</details>

### Prior Performance

This issue did not exist in pandas 1.2.5. The timing measured by the script with pandas 1.2.5 on my computer is about 30 milliseconds, while the timing in pandas 1.3.1, 1.3.2 and 1.3.3 is about 600 ms. Thus there has been a 20x increase in run time. This severely impacts the performance of our application. A section of our code that used to take about 30 seconds, now takes about 10 minutes (that code uses `pd.concat` a lot, which in turn calls `MultiIndex.equals`).

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #43549
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [ ] whatsnew entry

```
from dateutil import tz
import pandas as pd
from time import time

dates = pd.date_range('2010-01-01', periods=1000, tz=tz.tzutc())
index = pd.MultiIndex.from_product([range(100), dates])
index2 = index.copy()

%timeit index.equals(index2)
330 ms ± 6.44 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)  # <- master
1.79 ms ± 7.4 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)  # <- PR
```

### CODE REVIEW

File: pandas/core/indexes/multi.py [hunk 1]
```diff
@@ -3544,8 +3544,13 @@ def equals(self, other: object) -> bool:
             if len(self_values) == 0 and len(other_values) == 0:
                 continue
 
-            if not array_equivalent(self_values, other_values):
-                return False
+            if not isinstance(self_values, np.ndarray):
```
Comment (jreback) [inline pandas/core/indexes/multi.py:3547]: maybe should do this check *in* `array_equivalent`?

Comment (jbrockmendel) [reply pandas/core/indexes/multi.py:3547]: at some point probably; need to check if other usages will be affected by the semantics of EA.equals

**Comment by jreback** [comment]
prob worth backporting this, can you add a whatsnew note @jbrockmendel ping on green

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
@meeseeksdev backport 1.3.x

**Pull request merged**